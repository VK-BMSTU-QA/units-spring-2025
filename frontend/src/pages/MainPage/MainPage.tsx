import React, { useState } from 'react';

import type { Category } from '../../types';

import { useCurrentTime, useProducts } from '../../hooks';
import { Categories, ProductCard } from '../../components';
import { applyCategories, updateCategories } from '../../utils';

import './MainPage.css';

export const MainPage = () => {
    const products = useProducts();
    const currentTime = useCurrentTime();

    const [selectedCategories, setSelectedCategories] = useState<Category[]>(
        []
    );
    const onCategoryClick = (clickedCategory: Category) =>
        setSelectedCategories(
            updateCategories(selectedCategories, clickedCategory)
        );

    return (
        <div className="main-page">
            <h2 className="main-page__title">VK Маркет</h2>
            <h3 data-testid="time">{currentTime}</h3>
            <div className="main-page__parameters">
                <Categories
                    selectedCategories={selectedCategories}
                    onCategoryClick={onCategoryClick}
                />
            </div>
            {applyCategories(products, selectedCategories).map((product) => (
                <ProductCard key={product.id} {...product} />
            ))}
        </div>
    );
};
