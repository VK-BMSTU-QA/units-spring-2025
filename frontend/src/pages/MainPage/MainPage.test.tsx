import { render, screen, fireEvent } from '@testing-library/react';
import { MainPage } from './MainPage';
import { useProducts, useCurrentTime } from '../../hooks';
import '@testing-library/jest-dom';

jest.mock('../../hooks', () => ({
    useProducts: jest.fn(),
    useCurrentTime: jest.fn(),
}));

jest.mock('../../components/Categories', () => ({
    Categories: ({ onCategoryClick }: any) => (
        <div>
            <button data-testid="category-electronics" onClick={() => onCategoryClick('Электроника')}>Электроника</button>
            <button data-testid="category-clothing" onClick={() => onCategoryClick('Одежда')}>Одежда</button>
        </div>
    ),
}));

describe('MainPage', () => {
    beforeEach(() => {
        (useProducts as jest.Mock).mockReturnValue([
            { id: 1, name: 'Product 1', description: 'Смартфон', price: 1000, priceSymbol: '$', category: 'Электроника' },
            { id: 2, name: 'Product 2', description: 'Тёплая куртка', price: 200, priceSymbol: '$', category: 'Одежда' }
        ]);
        (useCurrentTime as jest.Mock).mockReturnValue('2025-03-13 12:00');
    });

    it('should render the title and current time', () => {
        render(<MainPage />);
        expect(screen.getByText('VK Маркет')).toBeInTheDocument();
        expect(screen.getByText('2025-03-13 12:00')).toBeInTheDocument();
    });

    it('should update selected categories when a category is clicked', () => {
        render(<MainPage />);
        fireEvent.click(screen.getByTestId('category-electronics'));
        fireEvent.click(screen.getByTestId('category-clothing'));

        expect(screen.getByTestId('category-electronics')).toBeInTheDocument();
        expect(screen.getByTestId('category-clothing')).toBeInTheDocument();
    });

    it('should apply selected categories and render products accordingly', () => {
        render(<MainPage />);
        fireEvent.click(screen.getByTestId('category-electronics'));

        const productCards = screen.getAllByText('Электроника', { selector: '.product-card__category' });
        expect(productCards).toHaveLength(1);
    });
});
