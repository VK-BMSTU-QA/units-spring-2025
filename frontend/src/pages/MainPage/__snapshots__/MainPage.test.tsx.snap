// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage test should render correctly 1`] = `
<DocumentFragment>
  <div
    class="main-page"
  >
    <h2
      class="main-page__title"
    >
      VK Маркет
    </h2>
    <h3>
      01:00:00
    </h3>
    <div
      class="main-page__parameters"
    >
      <div
        class="categories"
      >
        <div
          class="categories__badge"
        >
          Одежда
        </div>
        <div
          class="categories__badge"
        >
          Для дома
        </div>
        <div
          class="categories__badge"
        >
          Электроника
        </div>
      </div>
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          Принтер
        </h2>
        <div
          class="product-card__description"
        />
        <h1
          class="product-card__price"
        >
          100 $
        </h1>
        <div
          class="product-card__category"
        >
          Электроника
        </div>
      </div>
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          Костюм гуся
        </h2>
        <div
          class="product-card__description"
        />
        <h1
          class="product-card__price"
        >
          100 $
        </h1>
        <div
          class="product-card__category"
        >
          Одежда
        </div>
      </div>
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          Настольная лампа
        </h2>
        <div
          class="product-card__description"
        />
        <h1
          class="product-card__price"
        >
          100 $
        </h1>
        <div
          class="product-card__category"
        >
          Для дома
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
